### PR TITLE
Precise description of reload_dataloaders_every_n_epochs

### LINKED ISSUE #14093: Reload new data after one epoch doesnt work.
### To Reproduce

https://colab.research.google.com/drive/1V0LDGDjW_Ettv0q6fQXFtnd-j2aIURDm?usp=sharing
Just follow the boring model

### Expected behavior

tldr: The trainer should use or reload the data from the class npDataModule. The trainer doesnt do it because it never reaches the breakpoints there.

### Environment

- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
- PyTorch Lightning Version (e.g., 1.5.0):
- Lightning App Version (e.g., 0.5.2):
- PyTorch Version (e.g., 1.10):
- Python version (e.g., 3.9):
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Running environment of LightningApp (e.g. local, cloud):
- Any other relevant information:

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           11.3
* Packages:
	- lightning:         None
	- lightning_app:     None
	- numpy:             1.21.6
	- pyTorch_debug:     False
	- pyTorch_version:   1.12.0+cu113
	- pytorch-lightning: 1.7.0
	- tqdm:              4.64.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.13
	- version:           #1 SMP Sun Apr 24 10:03:06 PDT 2022

### Additional context

I would like to carry out Adversarial Defence Training. In the first epoch I load a data module and from the second epoch onwards I want to inject modified data into the training process at the beginning of each epoch.

I have three different approaches and they all don't work.

1.)  The feature of the   
```
reload_dataloaders_every_n_epochs
```
of the trainer class.
According to the [documentation](https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#reload-dataloaders-every-n-epochs)
I should use in my model (i.e. a LightningModule) the function  
``` train_dataloader(), val_dataloader() ``` etc., from which 
the trainer reloads the new dataloaders. This does not work.
2.) the usage of the functions
```python
trainer.reset_train_dataloader()
trainer.reset_val_dataloader()
```
Here it is not clear what is being loaded from. (I find the documentation very sparse). It turns out that a new ```training_dataloader``` has been loaded of type ```CombinedLoader```, which is not what I wanted. (```val_dataloader``` remains ```None```).
3) I manually set ```datamodule``` or ```train_dataloader``` etc. in the trainer object.
This does not work either. 

I am using the current version 1.7.
I know if I'm missing something (I can't find anything in the documentation), but I have a simple workflow from a data perspective (just new data) and I can't find people commenting about similar problems.
I'm not sure if it's a bug or I'm missing implicit assumptions.
I would still appreciate help as I have been sitting on this problem for many work days and am very desperate. Thank you very much!

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj

### PR DESCRIPTION
**Pull request by KralaBenjamin**
## What does this PR do?

As mentioned in the discussion I find the current description of reload_dataloaders_every_n_epochs confusing so I added more content to make the parameter more understandable. 

Fixes #14093

### Does your PR introduce any breaking changes? If yes, please list them.

None

### CODE REVIEW

File: docs/source-pytorch/common/trainer.rst [hunk 1]
```diff
@@ -1232,23 +1232,27 @@ reload_dataloaders_every_n_epochs
...
 
     # if a positive integer
     for epoch in epochs:
         if not epoch % reload_dataloaders_every_n_epochs:
-            train_loader = model.train_dataloader()
+            train_loader = current_datasource.train_dataloader()
         for batch in train_loader:
             ...
 
+Pseudo code applies also to the ``test_dataloader`` and the ``val_dataloader``.
```
Comment (rohitgr7) [inline docs/source-pytorch/common/trainer.rst:1254]:
```suggestion
The pseudocode also applies to the ``val_dataloader``.
```

File: docs/source-pytorch/common/trainer.rst [hunk 2]
```diff
@@ -1232,23 +1232,27 @@ reload_dataloaders_every_n_epochs
 
 |
 
-Set to a positive integer to reload dataloaders every n epochs.
+Set to a positive integer to reload dataloaders every n epochs from your currently used data source.
+DataSource can be a DataModule or a DataLightningModule.
```
Comment (rohitgr7) [inline docs/source-pytorch/common/trainer.rst:1236]:
```suggestion
Data-source can be a ``LightningModule`` or a ``LightningDataModule``.
```

**Review by rohitgr7 — APPROVED** [review]

**Comment by KralaBenjamin** [comment]
@rohitgr7  fix it.

File: docs/source-pytorch/common/trainer.rst [hunk 3]
```diff
@@ -1232,23 +1232,27 @@ reload_dataloaders_every_n_epochs
 
 |
 
-Set to a positive integer to reload dataloaders every n epochs.
+Set to a positive integer to reload dataloaders every n epochs from your currently used data source.
+DataSource can be a ``LightningModule`` or a ``LightningDataModule.
```
Comment (rohitgr7) [inline docs/source-pytorch/common/trainer.rst:1236]:
```suggestion
DataSource can be a ``LightningModule`` or a ``LightningDataModule``.
```

File: docs/source-pytorch/common/trainer.rst [hunk 4]
```diff
@@ -1232,23 +1232,27 @@ reload_dataloaders_every_n_epochs
...
+
 
 .. code-block:: python
 
     # if 0 (default)
-    train_loader = model.train_dataloader()
+    train_loader = current_datasource.train_dataloader()
```
Comment (awaelchli) [inline docs/source-pytorch/common/trainer.rst:1242]: I think this should stay concrete. You can put a comment line below like this
```py
train_loader = model.train_dataloader()
# or if using data module: datamodule.train_dataloader()
```

or similar.

**Comment by KralaBenjamin** [comment]
done

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**